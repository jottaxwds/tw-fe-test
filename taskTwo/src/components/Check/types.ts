export type CheckOption = {
  name: string;
  displayValue: string;
  value: string;
};

export type CheckChangeEventArg = {
  id: string;
  value: string;
  checked: boolean;
};

export type CheckGroupChangeEventArgs =
  | { id: string; value: string }
  | undefined;

export type CheckGroupProps = {
  label: string;
  options: CheckOption[];
  value: string;
  onCheckChange: (args?: CheckGroupChangeEventArgs) => void;
};
