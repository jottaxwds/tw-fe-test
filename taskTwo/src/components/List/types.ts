import { Meal } from "../../context/types";

export type ListProps = {
  items: Meal[];
  itemsPerPage?: number;
  isLoading?: boolean;
};
