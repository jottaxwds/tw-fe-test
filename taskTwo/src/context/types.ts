export type FilterType = "c" | "a" | "i" | "";
export type Filter = {
  type: FilterType;
  value: string;
};

export type Ingredient = {
  name: string;
  measure: string;
};

export type Meal = {
  id: string;
  area: string;
  category: string;
  ingredients: Ingredient[];
  instructions: string;
  thumbImg: string;
  title: string;
  source: string;
  youtube: string;
};

export type CachedMeal = {
  type: FilterType;
  value: string;
  meals: Meal[];
};

export type State = {
  cachedMeals: CachedMeal[];
  globalError: boolean;
  filter: Filter;
  randomMeal: Meal;
};

export enum ActionType {
  CLEAR_FILTER = "CLEAR_FILTER",
  SET_FILTER = "SET_FILTER",
  UPDATE_MEALS = "UPDATE_MEALS",
  UPDATE_RANDOM_MEAL = "UPDATE_RANDOM_MEAL",
}

export type ActionPayload = {
  meals?: Meal[];
  filter?: Filter;
  randomMeal?: Meal;
};

export interface Action {
  type: ActionType;
  payload?: ActionPayload;
}
