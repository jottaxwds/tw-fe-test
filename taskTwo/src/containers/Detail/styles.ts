import styled from "styled-components";
import theme from "../../theme";

export const DetailContainer = styled.div`
  background: ${theme.palette.lightBrown};
  height: 100%;
`;
export const Heading = styled.div`
  display: flex;
  flex-direction: row;
`;
export const Preview = styled.div<{ backgroundImageURL: string }>`
  display: flex;
  background-color: ${theme.palette.subdued};
  background-image: url("${({ backgroundImageURL }) => backgroundImageURL}");
  background-size: cover;
  background-position: center center;
  width: 300px;
`;
export const Title = styled.h2`
  color: ${theme.palette.accent};
  ${theme.typography.detailsHeading};
  margin: 0;
  margin-bottom: ${theme.spacing.small};
`;
export const Ingredients = styled.div`
  background: ${theme.palette.darkBrown};
  flex: 1;
  padding: ${theme.spacing.medium} ${theme.spacing.mediumLarge};
`;
export const IngredientsList = styled.ul`
  list-style: none;
  ${theme.typography.text};
  margin: 0;
  padding: 0;
`;
export const Ingredient = styled.li`
  margin-bottom: ${theme.spacing.xxSmall};
`;

export const Main = styled.main`
  background: ${theme.palette.lightBrown};
  ${theme.typography.text};
  display: flex;
  flex-direction: row;
`;

export const DummyAside = styled.aside`
  width: 300px;
`;

export const Article = styled.article`
  flex: 1;
  padding: ${theme.spacing.medium} ${theme.spacing.mediumLarge};
`;
