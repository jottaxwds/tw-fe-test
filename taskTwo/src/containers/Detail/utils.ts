export function getYouTubeIdFromUrl(url: string) {
  const youtubeVideoRegexp =
    /(?:https?:\/\/)?(?:www\.)?(?:youtube\.com|youtu\.be)\/(?:watch\?v=)?(.+)/;
  const match = url.match(youtubeVideoRegexp);
  return match ? match[1] : "";
}
