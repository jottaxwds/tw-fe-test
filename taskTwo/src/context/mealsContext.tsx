import React from "react";
import {
  Action,
  ActionPayload,
  ActionType,
  CachedMeal,
  Filter,
  Meal,
  State,
} from "./types";

const initialState = {
  cachedMeals: [],
  randomMeal: {} as Meal,
  globalError: false,
  filter: { type: "", value: "" } as Filter,
};

const MealsContext = React.createContext<[State, React.Dispatch<Action>]>([
  initialState,
  () => {},
]);

const filteredMealsAreCached = (cachedMeals: CachedMeal[], filter: Filter) =>
  cachedMeals.length &&
  cachedMeals.some(
    ({ value, type, meals }) =>
      value === filter.value && type === filter.type && meals.length
  );

function reducer(state: State, action: Action) {
  const { type, payload } = action;
  const {
    meals = [],
    randomMeal = {} as Meal,
    filter = { type: "" as const, value: "" } as Filter,
  } = payload as ActionPayload;

  switch (type) {
    case ActionType.SET_FILTER:
      return { ...state, filter };
    case ActionType.CLEAR_FILTER:
      return { ...state, filter: { type: "" as const, value: "" } };
    case ActionType.UPDATE_MEALS:
      const { cachedMeals, filter: stateFilter } = state;
      if (filteredMealsAreCached(cachedMeals, stateFilter)) {
        return { ...state };
      }
      return {
        ...state,
        cachedMeals: [
          ...state.cachedMeals,
          {
            type: state.filter.type,
            value: state.filter.value,
            meals,
          },
        ],
      };
    case ActionType.UPDATE_RANDOM_MEAL:
      return { ...state, randomMeal };
    default:
      return { ...state };
  }
}
function MealsProvider({ children }: { children: any }) {
  const [state, dispatch] = React.useReducer<React.Reducer<State, Action>>(
    reducer,
    {
      ...initialState,
    }
  );

  // eslint-disable-next-line react/jsx-no-constructed-context-values
  return (
    <MealsContext.Provider value={[state, dispatch]}>
      {children}
    </MealsContext.Provider>
  );
}

export { MealsContext, MealsProvider };
