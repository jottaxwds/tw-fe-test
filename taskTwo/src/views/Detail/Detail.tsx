import DetailContent from "../../containers/Detail/Detail";
import useMealDetail from "../../hooks/useMealDetail";

import * as S from "./styles";

const Detail = () => {
  const { isLoadingMealInfo, meal } = useMealDetail();
  return (
    <>
      <S.TopBar>Only Quality Food</S.TopBar>
      {isLoadingMealInfo ? (
        <S.Info data-testid={"detail-loading"}> Loading ... </S.Info>
      ) : (
        <></>
      )}
      {meal !== null ? (
        <DetailContent meal={meal} />
      ) : (
        <div data-testid={"detail-error"}>Error!</div>
      )}
    </>
  );
};

export default Detail;
