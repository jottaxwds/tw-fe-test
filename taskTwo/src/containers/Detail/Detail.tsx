import { snakeCase } from "lodash";
import React from "react";
import Youtube from "react-youtube";
import { Ingredient as IngredientType, Meal } from "../../context/types";
import * as S from "./styles";
import { getYouTubeIdFromUrl } from "./utils";

const youtubeVideoConfig = {
  height: "330",
  playerVars: {
    autoplay: 0,
  },
};

const Detail = ({ meal }: { meal: Meal }) => {
  const ingredientsList = React.useCallback(
    (ingredients: IngredientType[]) =>
      ingredients.length ? (
        ingredients
          .filter(
            ({ name, measure }) =>
              name &&
              name !== "" &&
              name !== undefined &&
              measure &&
              measure !== "" &&
              measure !== undefined
          )
          .map(({ name, measure }) => (
            <S.Ingredient key={snakeCase(`${name}-${measure}`)}>
              <strong>{measure}</strong>
              {` ${name}`}
            </S.Ingredient>
          ))
      ) : (
        <></>
      ),
    []
  );

  return (
    <>
      <S.Heading>
        <S.Preview backgroundImageURL={meal.thumbImg}></S.Preview>
        <S.Ingredients>
          <S.Title>{meal.title}</S.Title>
          <S.IngredientsList data-testid={"ingredients-list"}>
            {ingredientsList(meal.ingredients)}
          </S.IngredientsList>
        </S.Ingredients>
      </S.Heading>
      <S.Main>
        <S.DummyAside />
        <S.Article data-testid={"instructions"}>
          <Youtube
            videoId={getYouTubeIdFromUrl(meal.youtube)}
            opts={youtubeVideoConfig}
          />
          {meal.instructions.split(/\r\n|\r|\n/).map((p, index) => (
            <p key={`p-${meal.id}-${index}`}>{p}</p>
          ))}
        </S.Article>
      </S.Main>
    </>
  );
};

export default Detail;
