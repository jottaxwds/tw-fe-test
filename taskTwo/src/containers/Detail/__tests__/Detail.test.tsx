import { render, screen } from "@testing-library/react";
import { Meal } from "../../../context/types";
import { mealsExample } from "../../../testData";
import Detail from "../Detail";

describe("Detail container", () => {
  it("Should show ingredients list matching snapshot", () => {
    render(<Detail meal={mealsExample[0] as Meal} />);
    expect(screen.getByTestId("ingredients-list")).toMatchSnapshot();
  });

  it("Should show instructions matching snapshot", () => {
    render(<Detail meal={mealsExample[0] as Meal} />);
    expect(screen.getByTestId("instructions")).toMatchSnapshot();
  });

  it("Should show empty ingredients list if not provided in the meal", () => {
    render(<Detail meal={{ ...(mealsExample[0] as Meal), ingredients: [] }} />);
    expect(screen.getByTestId("ingredients-list")).toBeEmptyDOMElement();
  });
});
