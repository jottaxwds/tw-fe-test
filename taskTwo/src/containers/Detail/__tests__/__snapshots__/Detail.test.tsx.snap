// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Detail container Should show ingredients list matching snapshot 1`] = `
<ul
  class="sc-jTQDnj foyXlk"
  data-testid="ingredients-list"
>
  <li
    class="sc-gLLvby fSskKK"
  >
    <strong>
      3 tbsp
    </strong>
     Olive Oil
  </li>
  <li
    class="sc-gLLvby fSskKK"
  >
    <strong>
      2 chopped
    </strong>
     Onions
  </li>
  <li
    class="sc-gLLvby fSskKK"
  >
    <strong>
      2 sticks
    </strong>
     Celery
  </li>
  <li
    class="sc-gLLvby fSskKK"
  >
    <strong>
      300g
    </strong>
     Carrots
  </li>
  <li
    class="sc-gLLvby fSskKK"
  >
    <strong>
      500g
    </strong>
     Potatoes
  </li>
  <li
    class="sc-gLLvby fSskKK"
  >
    <strong>
      4
    </strong>
     Bay Leaf
  </li>
  <li
    class="sc-gLLvby fSskKK"
  >
    <strong>
      5 tblsp 
    </strong>
     Tomato Puree
  </li>
  <li
    class="sc-gLLvby fSskKK"
  >
    <strong>
      2 tblsp 
    </strong>
     Sugar
  </li>
  <li
    class="sc-gLLvby fSskKK"
  >
    <strong>
      2 tblsp 
    </strong>
     White Vinegar
  </li>
  <li
    class="sc-gLLvby fSskKK"
  >
    <strong>
      1½ kg
    </strong>
     Chopped Tomatoes
  </li>
  <li
    class="sc-gLLvby fSskKK"
  >
    <strong>
      500g
    </strong>
     Passata
  </li>
  <li
    class="sc-gLLvby fSskKK"
  >
    <strong>
      3
    </strong>
     Vegetable Stock Cube
  </li>
  <li
    class="sc-gLLvby fSskKK"
  >
    <strong>
      400ml
    </strong>
     Whole Milk
  </li>
</ul>
`;

exports[`Detail container Should show instructions matching snapshot 1`] = `
<article
  class="sc-qZqnM gFWgwv"
  data-testid="instructions"
>
  <div
    class=""
  >
    <div
      class=""
      id=""
    />
  </div>
  <p>
    Put the oil, onions, celery, carrots, potatoes and bay leaves in a big casserole dish, or two saucepans. Fry gently until the onions are softened – about 10-15 mins. Fill the kettle and boil it.
  </p>
  <p>
    Stir in the tomato purée, sugar, vinegar, chopped tomatoes and passata, then crumble in the stock cubes. Add 1 litre boiling water and bring to a simmer. Cover and simmer for 15 mins until the potato is tender, then remove the bay leaves. Purée with a stick blender (or ladle into a blender in batches) until very smooth. Season to taste and add a pinch more sugar if it needs it. The soup can now be cooled and chilled for up to 2 days, or frozen for up to 3 months.
  </p>
  <p>
    To serve, reheat the soup, stirring in the milk – try not to let it boil. Serve in small bowls with cheesy sausage rolls.
  </p>
</article>
`;
